compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/util/IrUtils.kt

```diff
@@ -465,46 +470,72 @@ fun irCall(
         ).apply {
             copyTypeAndValueArgumentsFrom(
                 call,
-                dispatchReceiverAsFirstArgument,
-                firstArgumentAsDispatchReceiver
+                receiversAsFirstArguments,
+                firstArgumentsAsReceivers
             )
         }
     }
 
+fun IrFunctionAccessExpression.copyTypeAndValueArgumentsFrom(
+    src: IrFunctionAccessExpression,
+    receiversAsFirstArguments: Boolean = false,
+    firstArgumentsAsReceivers: Boolean = false
+) = copyTypeAndValueArgumentsFrom(src, src.symbol.owner, symbol.owner, receiversAsFirstArguments, firstArgumentsAsReceivers)
+
+fun IrFunctionReference.copyTypeAndValueArgumentsFrom(
+    src: IrFunctionReference,
+    receiversAsFirstArguments: Boolean = false,
+    firstArgumentsAsReceivers: Boolean = false
+) = copyTypeAndValueArgumentsFrom(src, src.symbol.owner, symbol.owner, receiversAsFirstArguments, firstArgumentsAsReceivers)
+
 private fun IrMemberAccessExpression.copyTypeAndValueArgumentsFrom(
-    call: IrMemberAccessExpression,
-    dispatchReceiverAsFirstArgument: Boolean = false,
-    firstArgumentAsDispatchReceiver: Boolean = false
+    src: IrMemberAccessExpression,
+    srcFunction: IrFunction,
+    destFunction: IrFunction,
+    receiversAsFirstArguments: Boolean = false,
```

Comment:
Could you a bit explain why extension receiver should be moved to value parameters?
Seems both schemes should work properly.

NB: This renaming should also affect native code.

Reply:
Test that fails without this change: compiler/testData/codegen/box/traits/traitWithPrivateExtension.kt

Reply:
@gogabr Oh thanks. Now I caught the case. 
@punzki Please check that renaming and new logic not affect K/N

Reply:
That's right, that's the test case that fails.

@max-kammerer - This param is not used in K/N.

Reply:
Modified this so there is a separate set of parameters for the extension receiver. Looks like it is needed to address this failure: https://teamcity.jetbrains.com/viewLog.html?buildId=2383285&buildTypeId=Kotlin_dev_Compiler. InlineClassLowering creates functions where the dispatch receiver is moved to an argument but not the extension receiver.

Reply:
It could be unexpected behavior or bug in InlineClassLowering
@skuzmich @romanart Could you take a look?

Reply:
The relevant lines are here: https://github.com/JetBrains/kotlin/blob/d14859b44e1e1241a8e3c9d01b40b0b8175cf0c1/compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/InlineClassDeclarationLowering.kt#L267-L273. Extension receiver is copied as-is, but dispatch receiver is copied as a parameter. It seemed a little odd to me, but I don't know much about this lowering.

Reply:
@skuzmich seems extensionReceiverParameter also should be passes as value parameter

Reply:
NB: createStaticBodilessMethod seems could be merged with createStaticFunctionWithReceivers

Reply:
This was intentional. I thought keeping extension receiver would be less intrusive. Now I see that this is odd because conventional ordering receivers is changed. 

We can fix is here or in separate commit, but IrFunction and IrCall transformations have to be in sync either way.

Reply:
@punzki Would you take a look? Additional parameters for extension receiver become unnecessary

Reply:
@max-kammerer @skuzmich Something like this? (see latest commit)

Reply:
@punzki Yes! Many thanks!

Reply:
@punzki LGTM